skipruntime-ts/adapters/postgres/src/index.ts

```diff
@@ -0,0 +1,211 @@
+
+  if (!("col" in params["key"]) || typeof params["key"]["col"] != "string")
+    throw new Error(
+      "No key column specified for external Postgres data source",
+    );
+  const col: string = params["key"]["col"];
+
+  if (!("type" in params["key"]) || typeof params["key"]["type"] != "string")
+    throw new Error("No key type specified for external Postgres data source");
+  const type: string = params["key"]["type"];
+
+  let select;
+  switch (type) {
+    case "TEXT":
+      select = (table: string, key: string) =>
+        format("SELECT * FROM %I WHERE %I = %L;", table, col, key);
+      break;
+    case "BIGINT":
+      select = (table: string, key: string) => {
+        const keyNum = Number(key);
+        // Checks if key is a safe 64-bit integer
+        if (!Number.isSafeInteger(keyNum))
+          throw new Error("Invalid BIGINT key: " + key);
+        return format(
+          "SELECT * FROM %I WHERE %I = " + keyNum.toString() + ";",
+          table,
+          col,
+        );
+      };
+      break;
+    case "INTEGER":
+      select = (table: string, key: string) => {
+        const keyNum = Number(key);
+        // Checks if key is a safe 32-bit integer
+        if (
+          !Number.isSafeInteger(keyNum) ||
+          keyNum < min32bitInt ||
+          keyNum > max32bitInt
+        )
+          throw new Error("Invalid INTEGER key: " + key);
+        return format(
+          "SELECT * FROM %I WHERE %I = " + keyNum.toString() + ";",
+          table,
+          col,
+        );
+      };
+      break;
+    default:
+      throw new Error("Unsupported Postgres key type: " + type);
+  }
+  return { col, type, select };
+}
+
+/**
+ * An External Service wrapping a PostgreSQL database, exposing its tables as "resources" in the Skip runtime.
+ * Subscription `params` MUST include a field `key` identifying the table column that should be used as the key in the resulting collection (and its type, one of INTEGER, BIGINT, or TEXT).
+ */
+
+export class PostgresExternalService implements ExternalService {
+  private client: pg.Client;
+  private clientID: string;
+  private open_instances: Set<string> = new Set<string>();
+
+  isConnected(): boolean {
+    return "_connected" in this.client && !!this.client._connected;
+  }
+
+  constructor(db_config: {
+    host: string;
+    port: number;
+    database: string;
+    user: string;
+    password: string;
+  }) {
+    //generate random client ID for PostgreSQL notifications
+    this.clientID = "skip_pg_client_" + Math.random().toString(36).slice(2);
+
+    this.client = new pg.Client(db_config);
+    this.client.connect().then(
+      () => this.client.query(format(`LISTEN %I;`, this.clientID)),
+      () => {
+        throw new Error(
+          "Error connecting to Postgres at " + JSON.stringify(db_config),
+        );
+      },
+    );
+  }
+
+  subscribe(
+    instance: string,
+    table: string,
+    params: Json,
```

Comment:
Would it make sense / be possible to refine this to something like `Json & {key: "INTEGER" | "BIGINT" | "TEXT"}`?